Add support for generation of repomd.xml.asc files

PaulSD opened the pull request:
fixes #3055
https://pulp.plan.io/issues/3055

dkliban commented:
ok test

dkliban commented:
@PaulSD, Thank you so much for putting together this pull request. I was very glad to see that you created an issue in Redmine also. 

Since this is a new feature for Pulp, I switched issue 3055 to be a Story. Could you please update the  subject to something like "As a user, I can publish RPM repository with repomd.xml.asc file"

In the description of the ticket please outline how this feature will be used by a user. Will there be an additional REST api parameter that can be passed in when publishing? or is this going to be a distributor config option? Will this feature be available from the CLI? 

Thank you again! I look forward to having this feature being in Pulp.

PaulSD commented:
Done.  Let me know if there is anything else I should do.  Thanks!

bmbouter commented:
@PaulSD btw, thank you so much for submitting this PR!

bmbouter commented on plugins/pulp_rpm/plugins/distributors/yum/metadata/repomd.py:
```diff
@@ -32,6 +36,18 @@ def __exit__(self, exc_type, exc_val, exc_tb):
+                                    stderr=subprocess.PIPE)
+            stdout, stderr = pipe.communicate()
+            if pipe.returncode != 0:
+                raise PulpCodedException(error_codes.RPM0001, command='gpg', stdout=stdout,
```
Would you want to create a new coded exception for an error encountered during signing of metadata? It would be a new entry here:  https://github.com/pulp/pulp_rpm/blob/master/plugins/pulp_rpm/plugins/error_codes.py#L38

I think that would be good because this is a specific place where an error could occur. Add it if you want; it's an idea not a required change.

dkliban commented:
ok test

PaulSD commented:
I updated the release notes, and adjusted the exception to include the full command line in the message.
I could add a new coded exception if necessary, but this is really indicating a generic problem external to pulp (as opposed to a specific problem within pulp), so it actually seems like RPM0001 is appropriate here, and should already include enough info for the user to identify the root cause even without a separate coded exception.

Three thoughts/questions I might want to explore some more:
1) I don't see much reason to enable/disable this per-repo ... The user is either going to want this on for all repos or off for all repos.  However, I didn't see any good way to set/get a global config value that is used by all configured Yum distributors.  Am I missing something, or is my current per-repo config really the best option here?
2) I am currently using a single gpgkey file to hold both the metadata signing key and the RPM signing key.  It would probably make sense to allow two separate files to be used for this.  It looks like the gpgkey config parameter could probably hold an array of files, or possibly multiple newline-separated filenames in a string, or perhaps a second parameter could be added for the metadata key file.  However, I don't have a full picture of how/where gpgkey gets passed around and used, so I'm not confident that this will work.  Any ideas/suggestions here?
3) I'm using Satellite, and yum seems to automagically grab gpg keys from Satellite.  Does the gpgkey config parameter on Yum distributors in Pulp have anything to do with this, or is Satellite handling this independently?  Would Pulp have anything to do with this if I was using Pulp without Satellite?

mhrivnak commented:
> However, I didn't see any good way to set/get a global config value that is used by all configured Yum distributors. Am I missing something, or is my current per-repo config really the best option here?

That is available in /etc/pulp/server/plugins.conf.d/ by creating a json file named "yum_distributor.json", and you can put any values in it that you would put in the json structure used to express the distributor's config to the REST API.

That said, I wonder if signing keys might be special enough to be expressed elsewhere. I could see configuring a single signing key in server.conf for example that anything could use. We could concoct fancier use cases of people signing different publications differently for various reasons, but enabling it with one key to at least prove where the metadata came from would go a long way.

> I am currently using a single gpgkey file to hold both the metadata signing key and the RPM signing key. It would probably make sense to allow two separate files to be used for this. It looks like the gpgkey config parameter could probably hold an array of files, or possibly multiple newline-separated filenames in a string, or perhaps a second parameter could be added for the metadata key file. However, I don't have a full picture of how/where gpgkey gets passed around and used, so I'm not confident that this will work. Any ideas/suggestions here?

You're right that a lot of users will be managing third-party content with Pulp, so they won't have the private key that was used to sign the RPMs. 

If Pulp's signing key were configured in server.conf, then that would give you a standard way to access it in-code. In any case, we definitely need to store it somewhere that can be kept very secure.

> I'm using Satellite, and yum seems to automagically grab gpg keys from Satellite. Does the gpgkey config parameter on Yum distributors in Pulp have anything to do with this, or is Satellite handling this independently? Would Pulp have anything to do with this if I was using Pulp without Satellite?

I'm pretty sure katello-agent, a service running on each managed machine, puts the public part of the GPG key in place and configures yum to use it. katello-agent is closely related to pulp-agent and consists of mostly the same code. @jortel can remind us of the details, but if you wanted to use pulp-agent, it could likely do this for you.

That said, host management will not be a part of Pulp 3, so it's probably worth finding a different way to solve this. Whatever config management solution you're using is likely the best place to start for how to setup yum repos with the correct GPG key.

beav commented on docs/tech-reference/yum-plugins.rst:
```diff
@@ -596,9 +596,10 @@ Optional Configuration Parameters
- GPG key used to sign RPMs in this repository. This key will be made available
- to consumers to use in verifying content in the repository. The value to this
- option must be the full path to the GPG key file.
+ File containing GPG keys used to sign RPMs and metadata in this repository.
```
This is a little confusing. The "File containing GPG keys used to sign RPMs" is just used for clients to validate GPG signatures, but you can read it differently and it reads as though Pulp is doing actual GPG signing of RPMs. However, the metadata is actually signed by Pulp in this PR.

bmbouter commented:
@PaulSD Thanks again for this great contribution, the issue writeup, and revisions. I think merging this now and working on a better solution for Pulp3 would be great. We could discuss a Pulp3 option on [the pulp-dev mailing list](https://www.redhat.com/mailman/listinfo/pulp-dev) if you want. Overall I think having gpg signing as a feature like this one in Pulp3's core and having the Pulp3 Plugin API have a signing interface so that plugins like RPM can ask core to sign something for them would be ideal. It's an idea...

Would you be interested/able to give a demo of this feature? If so you can send me a recording or do a live one. We have the call for demo presenters on pulp-list also.

Also we send our first time contributors to Pulp a Pulp t-shirt and stickers. If you PM me on IRC ( bmbouter ) we can get you those.

PaulSD commented:
Sorry, I got busy with other things and this fell off my radar for a couple weeks.

> That is available in /etc/pulp/server/plugins.conf.d/ by creating a json file named "yum_distributor.json", and you can put any values in it that you would put in the json structure used to express the distributor's config to the REST API.

Excellent! I will add this to the docs and open another PR for the doc updates.

> That said, I wonder if signing keys might be special enough to be expressed elsewhere. I could see configuring a single signing key in server.conf for example that anything could use.

I think I already have this covered...
Both the `gpg` command and the python-gnupg library require private GPG keys to be stored in GPG keyring file ; You cannot use private keys from a configuration file or database or anything like that unless you write the keys out to a GPG keyring file before using them.  So, rather than add a configuration value to Pulp for the private GPG key (which we would then have to write out to a GPG keyring file before use), I simply require the user to create a GPG keyring file on their own (independently from Pulp), and then configure Pulp to use that GPG keyring file (either by placing the file in the default ~/.gnupg/ path or by setting a $GNUPGHOME environment variable).  The only configuration related to this behavior that is actually in Pulp is the `gpg_sign_metadata` flag which determines whether signing will occur ; Everything else is configured externally to Pulp.

The `gpgkey` configuration in Pulp is used to provide public GPG keys to clients so they can verify the metadata and/or RPMs in the repository, which is only tangentially related to the signing functionality added by this PR.

> We could concoct fancier use cases of people signing different publications differently for various reasons, but enabling it with one key to at least prove where the metadata came from would go a long way.

Yeah, I did not bother implementing support for choosing between multiple GPG signing keys.  Currently, if the GPG keyring contains multiple signing keys, the `gpg` command will automatically choose one to use (I believe it uses whatever key is newest), and it will end up using the same key for all repos.  We could add a configuration option to allow different signing keys to be used for different repos (by simply passing the appropriate key ID to the `gpg` command during signing), but I couldn't think of any use cases where this behavior would actually be needed when signing metadata, so I didn't bother with this ... It might be needed when signing RPMs, but I don't think it is needed for metadata.

> I'm pretty sure katello-agent, a service running on each managed machine, puts the public part of the GPG key in place and configures yum to use it. katello-agent is closely related to pulp-agent and consists of mostly the same code. jortel can remind us of the details, but if you wanted to use pulp-agent, it could likely do this for you.
> That said, host management will not be a part of Pulp 3, so it's probably worth finding a different way to solve this. Whatever config management solution you're using is likely the best place to start for how to setup yum repos with the correct GPG key.

I think this answers both of my second two questions from my earlier post.  As long as Satellite handles this separately from Pulp, I don't need to worry too much about the behavior of the `gpgkey` configuration value in Pulp, and Katello already provides an appropriate "different way to solve this".

> This is a little confusing. The "File containing GPG keys used to sign RPMs" is just used for clients to validate GPG signatures, but you can read it differently and it reads as though Pulp is doing actual GPG signing of RPMs. However, the metadata is actually signed by Pulp in this PR.

Agreed, I will include this in my new PR for doc updates.

kfiresmith commented:
This is actually wicked good timing to include this, because the RHEL7 STIG (specifically, RHEL-07-020070) requires this security feature.
